### PR TITLE
refactor: add ProxmoxClient to handle pve api interactions

### PR DESCRIPTION
**Pull request by peschmae**
This is just a draft, how I would implement #195 

All unit tests are green again, and all calls to the Proxmox API are wrapped in the `ProxmoxClient` class.

I'm considering this to be WIP, because I'm looking for some feedback, and I would like to add a unit test for `prometheuspvesd.discovery.Discovery.propagate`

### CODE REVIEW

File: prometheuspvesd/test/unit/test_discovery.py [hunk 1]
```diff
@@ -97,8 +87,8 @@ def test_validate_ip(discovery, addresses):
         assert not discovery._validate_ip(address)
 
 
-def test_get_ip_addresses(mocker, discovery, networks):
-    discovery.client.get.side_effect = lambda *args: get_mock(networks, *args)
+def test_get_ip_addresses(discovery, networks):
+    discovery.client.get_network_interfaces = MagicMock(return_value=networks)
```
Comment (xoxys) [inline prometheuspvesd/test/unit/test_discovery.py:91]: Can we use the mocker instead?

`mocker.patch.object(ProxmoxClient, "get_network_interfaces", return_value=networks)`

**Review by xoxys — CHANGES_REQUESTED** [review]
Thanks a lot. Just a small one so far, besides that, it looks pretty good to me already. Will take a closer look tomorrow.

**Comment by xoxys** [comment]
What is the intention of the "Change IP Fixtures to documentation range" commits? What documentation range?

**Comment by peschmae** [comment]
> What is the intention of the "Change IP Fixtures to documentation range" commits? What documentation range?

According to IANA, there are some reserved blocks, that should be used for documentation purposes. https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml (or IPv6: https://www.iana.org/assignments/iana-ipv6-special-registry/iana-ipv6-special-registry.xhtml )

The idea is, to avoid using IP ranges, that are used for real applications, in documentation and also CI setups.

**Comment by peschmae** [comment]
This is ready to be merged in my opinion.

I've stumbled across an issue with the ivp6 regex, and also fix that in this pull request, even though it would warant it's own fix.

The tests are all adjusted to use the new proxmox client class, and a test for `prometheuspvesd.discovery.Discovery.propagate` was added

File: prometheuspvesd/test/fixtures/fixtures.py [hunk 2]
```diff
@@ -222,12 +240,40 @@ def qemus():
...
+            "102.example.com",
+        "description":
+            '{"groups": "test-group"}',
+        "net0":
+            "virtio=D8-85-75-47-2E-8D,bridge=vmbr122,ip=192.0.2.25,ip=2001:db8::6666:7777:8888",
+        "cpu":
+            2,
+        "cores":
+            2
+    }
```
Comment (xoxys) [inline prometheuspvesd/test/fixtures/fixtures.py:256]:
```suggestion
    return {
        "name": "102.example.com",
        "description": '{"groups": "test-group"}',
        "net0": "virtio=D8-85-75-47-2E-8D,bridge=vmbr122,ip=192.0.2.25,ip=2001:db8::6666:7777:8888",  # noqa
        "cores": 2
    }
```

**Review by xoxys — CHANGES_REQUESTED** [review]
Could you rebase to main and commit the reformatted dict? I've adjusted the yapf config a bit to avoid line breaks.

File: prometheuspvesd/test/fixtures/fixtures.py (hunk 2 above)
Comment (peschmae) [reply prometheuspvesd/test/fixtures/fixtures.py:256]: I can do that, but then flake 8 isn´t happy with the net0 line as it is too long :/
But I can shorten the ipv6 address a bit, and keep it valid

Comment (xoxys) [reply prometheuspvesd/test/fixtures/fixtures.py]: Thats whats the `# noqa´ is for :)

**Review by xoxys — APPROVED** [review]

**Comment by xoxys** [comment]
Good work, thanks!

**Pull request merged**